### ISSUE TITLE
Continous restarts on invalid job configuration

### ISSUE THREAD

**Issue by sevensolutions** [hashicorp/nomad#17953]
### Nomad version
Nomad v1.5.6
BuildDate 2023-05-19T18:26:13Z
Revision 8af70885c02ab921dedbdf6bc406a1e886866f80

### Operating system and Environment details
Windows 10

### Issue
When running a job with an erroneous job configuration the task gets restarted infinitely although the last event message is "Not Restarting | Error was unrecoverable".
If the configuration is wrong the task will not start, no matter how often the scheduler will try, so i think it should stop creating new allocations.

### Reproduction steps
Run the job shown below. It contains the invalid iam_not_supported property.

#### Expected Result
On a configuration error the scheduler should stop creating new allocations.

#### Actual Result

![grafik](https://github.com/hashicorp/nomad/assets/84123899/ec66180a-de5e-4a79-9b0d-260243393709)

### Job file (if appropriate)

```hcl
job "test" {
  task "webservice" {
    driver = "raw_exec"

    config {
      command = "my-binary"
      args    = ["-flag", "1"]
      
      iam_not_supported = true
    }
  }
}
```

**Comment by schmichael** [comment]
> When running a job with an erroneous job configuration the task gets restarted infinitely although the last event message is "Not Restarting | Error was unrecoverable".

You're right this is very confusing! It's _technically_ accurate though, but only if know Nomad's particular distinction between [`restart`](https://developer.hashicorp.com/nomad/docs/job-specification/restart) and [`reschedule`](https://developer.hashicorp.com/nomad/docs/job-specification/reschedule). Nomad is *not* restarting the invalid task, but it *is rescheduling it!*

The root cause is the same as #5729 and #9287: the `task.config` block can only be validated by task drivers. Task drivers only run on Clients, so Servers cannot validate them ahead of time.

The fact that `reschedule.unlimited = true` is the default for `service` jobs makes this even worse because as you noted we will attempt to reschedule jobs that cannot run forever. However we can't safely change that default just for this unfortunate missing feature.

**Comment by sevensolutions** [comment]
Thank you for your answer, i understand that.
But there are two additional questions:
1. Task drivers already expose the task-config's schema. Can't we compare this before scheduling the task? Or is this task schema definition optional?
2. I'am writing my own task driver and i'am also validating the config before starting a task, but which response should i return in order to make the scheduler stop from re-scheduling the task?
The StartTask Result is defined like this:
```
message StartTaskResponse {
    enum Result {
        SUCCESS = 0;
        RETRY = 1;
        FATAL = 2;
    }
}
```
But even if i return *Fatal* the scheduler will re-schedule it. What is the difference between *Retry* and *Fatal*?
Which Result should the task driver return in case it detects a configuration error where it doesn't make sense to re-start or re-schedule the task?

**Comment by schmichael** [comment]
> 1. Task drivers already expose the task-config's schema. Can't we compare this before scheduling the task? Or is this task schema definition optional?

The problem is that there are two types of Nomad Agents: Servers (schedulers) and Clients (runs allocations, task drivers, etc). Servers do not load task driver plugins and therefore cannot parse and validate `task.config` prior to scheduling.

Servers _should_ just load plugins and validate `task.config`, but there are multiple complications:

1. Whenever you upgrade a task driver plugin, you must upgrade your servers as well.
2. Running multiple versions of a task driver plugin on different Clients is fine, but then which version do you use for validation on the Server?
3. Task driver plugin configuration parameters can impact whether a particular `task.config` value should be considered valid or not.

So I think there's 2 approaches we could take, although they are not mutually exclusive:

1. Load plugins on Servers and validate `task.config` (just like we did before task driver plugins existed).
2. Have Clients register their task driver plugin schemas with Servers. Task Driver Plugins must expose their `task.config` _schema_ to the Nomad Client, and this schema could at least validate the _structure_ of `task.config` if not the particular values.

The benefit of 2 is that you don't have to touch Servers to change plugins.

> 2. I'm writing my own task driver and i'am also validating the config before starting a task, but which response should i return in order to make the scheduler stop from re-scheduling the task?

You can't. Rescheduling is fully controlled by the jobspec author and is handled by the Server/scheduler. There's nothing a plugin can do to impact it until we implement 1 or 2 above (or change rescheduling behavior somehow).

It's an unfortunate situation!

**Comment by sevensolutions** [comment]
Ah i thought the clients will send the task driver's config schema to the servers but yes, you're right, i didn't think about the conflicting versions.
Thanks for the detailed explanation.
